river/metrics/cluster/sd_validation.py

```diff
@@ -0,0 +1,175 @@
+
+    * Scat(NC) stands for the scattering within clusters,
+
+    * Dis(NC) stands for the dispersion between clusters.
+
+    Like DB and SB, SD measures the compactness with variance of clustered objects and separation
+    with distnace between cluster centers, but uses them in a different way. The smaller the value
+    of SD, the better.
+
+    In the original formula for SD validation index, the ratio between the maximum and the actual
+    number of clusters is taken into account. However, due to the fact that metrics are update live
+    along with the data stream, this ratio will be automatically set to default as 1.
+
+    Examples
+    --------
+
+    >>> from river import cluster
+    >>> from river import stream
+    >>> from river import metrics
+
+    >>> X = [
+    ...     [1, 2],
+    ...     [1, 4],
+    ...     [1, 0],
+    ...     [4, 2],
+    ...     [4, 4],
+    ...     [4, 0],
+    ...     [-2, 2],
+    ...     [-2, 4],
+    ...     [-2, 0]
+    ... ]
+
+    >>> k_means = cluster.KMeans(n_clusters=3, halflife=0.4, sigma=3, seed=0)
+    >>> metric = metrics.cluster.SD()
+
+    >>> for x, _ in stream.iter_array(X):
+    ...     k_means = k_means.learn_one(x)
+    ...     y_pred = k_means.predict_one(x)
+    ...     metric = metric.update(x, y_pred, k_means.centers)
+
+    >>> metric
+    SD: 4.332702
+
+    References
+    ----------
+    [^1]: Halkidi, M., Vazirgiannis, M., & Batistakis, Y. (2000). Quality Scheme Assessment in the
+          Clustering Process. Principles Of Data Mining And Knowledge Discovery, 265-276.
+          DOI: 10.1007/3-540-45372-5_26
+
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._center_all_points = {}
+        self._overall_variance = {}
+        self._cluster_variance = {}
+        self._centers = {}
+        self._initialized = False
+
+    @staticmethod
+    def _calculate_dispersion_nc(centers):
+        min_distance_clusters = math.inf
+        max_distance_clusters = -math.inf
+        sum_inverse_distances = 0
+
+        n_clusters = len(centers)
+
+        for i in range(n_clusters):
+            for j in range(i + 1, n_clusters):
+                distance_ij = math.sqrt(
+                    utils.math.minkowski_distance(centers[i], centers[j], 2)
+                )
+                if distance_ij > max_distance_clusters:
+                    max_distance_clusters = distance_ij
+                if distance_ij < min_distance_clusters:
+                    min_distance_clusters = distance_ij
+                sum_inverse_distances += 1 / distance_ij
+
+        return max_distance_clusters / min_distance_clusters * sum_inverse_distances
+
+    @staticmethod
+    def _norm(x):
+        origin = {i: 0 for i in x}
+        return math.sqrt(utils.math.minkowski_distance(x, origin, 2))
+
+    def update(self, x, y_pred, centers, sample_weight=1.0):
+
+        if not self._initialized:
+            self._center_all_points = self._overall_variance = {
```

Comment:
Should both the variables point to the same region in memory?

If not, the initialization should be separate. Maybe other cases similar to this one escaped my attention.

Besides, is it possible to use `stats.Var` to keep the running variance? From my personal experience (and plenty of text-book references :P), it's always better to use the Welford algorithm whenever it's possible.

Reply:
You are correct! Indeed, I previously thought that variance of whole dictionary can not be calculated, but it seems that calculating variance of each dimension using `stats.Var` is still feasible!